### PR TITLE
Gramlib: improved support for multi-token LIST0/LIST1 separators and related minor changes

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

This PR does the following:
- introduce a new gramlib/camlp5 primitive symbol `tokens` for list of tokens; this has the following consequences:
  - list separators made of several tokens can be treated as first-order citizens: there is no more need to encode them as a rule of the form `["foo" "bar" -> ()]`
  - comparison of such separators is now possible (rules are arbitrary code and are always considered distinct by default); in particular, in the following example:
    ```coq
     Notation "[ a + + .. + + c | d ]" := (cons a .. (cons c d) ..) (a at level 10).
     Notation "[ a + + .. + + c ]" := (cons a .. (cons c nil) ..) (a at level 10).
     Check [0 + + 1 | nil].  (* now works *)
     Check [0 + + 1].
     ```
    the parser is now able to recognize that the separators are the same and to factorize the rules
  - the API for `LIST0 SEP` and `LIST1 SEP` can be cleaned: it is possible to give to the action associated to the separators the type `unit` even when the separator is a single token; this is a clean-up in that the action, even if not formally of type `unit`, was silently discarded
- incidentally, we support the `mlg` shortcut `[ symbs ]` for `[ symbs -> { () } ]` in `coqpp`: this notation was used by the gramlib printer but not accepted by `coqpp` (@ppedrot, is it ok for you?) (removed: redundant with #11910)
- add supports for terminals of different types (useful to treat both numbers and strings as terminals even if their associated tokens have different types)
- a few minor changes (moved to #14429)
  - moving `token_text` from `cLexer.ml` to `tok.ml` which looks more natural
  - introducing symmetry between `token_using` and `token_removing` (even if the effect of `token_removing` is overriden by state unfreezing) [incidentally, there is a problem with the state of keywords in plugins which need to be fixed]
  - add symmetry in the interchangeability of `IDENT` and `KEYWORD` when the corresponding identifier is a keyword

- [X] Added / updated test-suite
- [ ] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Review by jfehrle — COMMENTED** [review]
An initial thought: Have you considered how this will affect. doc_grammar and how Sphinx generates documentaron?  I believe you need to run "make doc_gram_rsts" to compile doc_grammar; it may be broken.

**Comment by herbelin** [comment]
@jfehrle: OK, I'll test this at some time. It can still change anyway.

**Comment by jfehrle** [comment]
Also, I have trouble imagining where in the grammar this would feature be helpful. 
It seems like it would be wordy.  Do you have a concrete example where you'd like to use this in the grammar?

Another consideration is that we have special rendering in the documentation for LIST1/LIST0 separators, e.g. the following is for LIST0 ltac_expr SEP "|".  @Zimmi48 and I agreed that this doesn't work for multicharacter separators (let alone multitoken separators).  Anywhere in the grammar where this is used, we would have to add editing commands for doc_grammar to put it in a form that will render well in the doc.

image

**Comment by herbelin** [comment]
>  Do you have a concrete example where you'd like to use this in the grammar?

In theory, vernac-level notations accept separators made of any sequence of tokens. In real life, it seems that recursive notations mostly have no separators, or `,`, or `;`, or `|`, or `!`, or `;;` (only the test-suite has multi-tokens separators).

So, if my motivation was to have gramlib `LIST1SEP` matching more closely what `Notation` supports, it seems indeed in practice that there is no concrete interest for it.

Note that from the point of view of `Print Grammar`, there is no change. A multi-token separator was printed `LIST1 (term LEVEL "200") SEP [ "+"; "+" ]` and it is still printed the same. As for parsing, there is no way to use the compact multitoken representation for `SEP`. It has to be written `[ list of tokens ]` which in turn is interpreted as a rule `[ list of tokens -> () ]`.

So, in summary, the support for multi-token in `LIST0SEP`/`LIST1SEP` seems overkill in practice, but it simplifies the `Notation` -> `LIST0SEP`/`LIST1SEP` translation and the new token `tokens` allow to give a better-typed API to `list0sep`/`list1sep`. Does it impact `doc_grammar` and the reference manual? I would say no, as long as `doc_grammar` relies on the `coqpp` parser which is unchanged by the PR.

So, if confirmed that there is no impact besides on gramlib itself, we can still wonder whether the small advantages balance the burden of a change of habits. Maybe can we take some time to ponder it.

**Comment by jfehrle** [comment]
> Does it impact `doc_grammar` and the reference manual? I would say no, as long as `doc_grammar` relies on the `coqpp` parser which is unchanged by the PR.

I agree, if used only in notations it would not affect `doc_grammar`.

**Comment by herbelin** [comment]
@parsing-mainteners: this is ready on my side.

**Comment by herbelin** [comment]
I mispelled @coq/parsing-maintainers, so sending again.

**Comment by herbelin** [comment]
That would be good to finally being able to merge the #14071, #13664, #13353 dependent chain of PRs. They bring small enhancements but, before all, they allow to continue building on cleaner bases.

File: gramlib/grammar.ml [hunk 1]
```diff
@@ -153,6 +156,7 @@ and ('trecs, 'trecp, 'a) ty_rec_level = {
 
 and ('self, 'trec, 'a) ty_symbol =
 | Stoken : 'c pattern -> ('self, norec, 'c) ty_symbol
+| Stokens : ty_pattern list -> ('self, norec, unit) ty_symbol
```
Comment (ppedrot) [inline gramlib/grammar.ml:157]: It looks a bit dangerous to me to have a redundancy between `Stoken` and `Stokens`. Maybe we should merge both nodes in a way that ensures that `Stokens` of a singleton has the same typing rule as before.

File: coqpp/coqpp_parse.mly [hunk 2]
```diff
@@ -391,6 +391,9 @@ rules:
 rule:
 | symbols_opt ARROW CODE
   { { gprod_symbs = $1; gprod_body = $3; } }
+| symbols
```
Comment (ppedrot) [inline coqpp/coqpp_parse.mly:394]: This one is a *cavalier législatif*, and since we have another PR containing it I don't think it should go in this one. (Regardless, I think this feature is making the code *less* robust since you can easily get into silent rule mismatch when rebasing code.)

**Review by ppedrot — COMMENTED** [review]
This PR is making many changes at one, and while I am not strongly opposed to merging it as-is, I think it'd be better for posterity to split it into self-contained chunks.

File: coqpp/coqpp_parse.mly (hunk 2 above)
Comment (herbelin) [reply coqpp/coqpp_parse.mly:394]: Beside being convenient, this was also to restore a symmetry with the output of `Print Grammar`. Would you prefer we modify the printing to adapt it to coqpp?

File: gramlib/grammar.ml (hunk 1 above)
Comment (herbelin) [reply gramlib/grammar.ml:157]: The redundancy is not nice indeed, but they have different types what justifies to have both. I agree that it is a compromise.

**Comment by herbelin** [comment]
I split the PR. The minor cleanup went to #14429 and I removed the coqpp part since it is already in #11910. Both this PR and  #14429 are standalone.

File: gramlib/grammar.ml (hunk 1 above)
Comment (ppedrot) [reply gramlib/grammar.ml:166]: You could type Stokens with a product type of the subrules so that when they have a length of one, they have the same rules a before. I don't think it is necessary for this PR though, it can always be refined later.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
I will try to write a patch factorizing Stoken with Stokens soon.
